### PR TITLE
allow negative freq strings

### PR DESCRIPTION
**Pull request by mathause**
- [ ] Closes #xxxx
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

This allows negative freq strings as discussed in https://github.com/pydata/xarray/pull/8627#issuecomment-1905981660 Deciding which tests to update was not easy.

The pandas `_generate_range` function was moved to https://github.com/pandas-dev/pandas/blob/3c96b8ff6d399fbec8d4d533e8e8618c592bb64b/pandas/core/arrays/datetimes.py#L2725 They no longer `rollback` the `end`. I had to remove this as well such that the following are eqivalent:

```python
xr.date_range("2001", "2000", freq="-1YE", calendar="noleap")
pd.date_range("2001", "2000", freq="-1YE")
```

I am slightly nervous about this but all the tests still pass...

Once again cc @spencerkclark

### CODE REVIEW

**Comment by dcherian** [comment]
Just so I understand, this is not backwards-incompatible because we did not allow negative frequency strings earlier?

**Comment by mathause** [comment]
> Just so I understand, this is not backwards-incompatible because we did not allow negative frequency strings earlier?

Yes that's correct.

**Review by spencerkclark — COMMENTED** [review]
This looks great @mathause—thanks!  

I think the tests you updated look pretty solid.  In my testing it seems like it just works, but maybe one more you could add would be an explicit test that calling `infer_freq` on a decreasing `CFTimeIndex` returns a negative frequency?

File: xarray/tests/test_cftime_offsets.py [hunk 1]
```diff
@@ -1541,3 +1602,26 @@ def test_to_offset_deprecation_warning(freq):
...
+
+@pytest.mark.parametrize("start", ("2000", "2001"))
+@pytest.mark.parametrize("end", ("2000", "2001"))
+@pytest.mark.parametrize("freq", ("M", "-1M", "Y", "-1Y"))
+def test_cftime_range_end_same_len_pandas(start, end, freq):
+    result = date_range(start, end, freq=freq, calendar="standard", use_cftime=True)
+    expected = date_range(start, end, freq=freq, use_cftime=False)
+
+    # end of month/ year is not convertable to datetimeindex
+    assert result.size == expected.size
```
Comment (spencerkclark) [inline xarray/tests/test_cftime_offsets.py:1627]: I think we can check for equality now, similar to the test above:
```suggestion
    result = date_range(start, end, freq=freq, calendar="standard", use_cftime=True)
    result = result.to_datetimeindex()
    expected = date_range(start, end, freq=freq, use_cftime=False)

    np.testing.assert_array_equal(result, expected)
```

**Review by spencerkclark — APPROVED** [review]

File: xarray/tests/test_cftime_offsets.py (hunk 1 above)
Comment (spencerkclark) [reply xarray/tests/test_cftime_offsets.py:1627]: Ultimately I suppose this test could now be merged with the one above—i.e. just merge the frequency strings.

Comment (mathause) [reply xarray/tests/test_cftime_offsets.py:1627]: of course - thanks
